### PR TITLE
CAS: Create a new NodeSchema/SchemaPool

### PR DESCRIPTION
**Pull request by dexonsmith**
Generalize the casobjectformats::SchemaBase and
casobjectformats::SchemaPool to cas::NodeSchema and cas::SchemaPool,
respectively, renaming the old things to ObjectFormatSchema{Base,Pool}
and connecting them via an open RTTI hierarchy using RTTIExtends.

This is groundwork for modelling trees as a node schema.

### CODE REVIEW

**Comment by dexonsmith** [comment]
@swift-ci test

**Comment by dexonsmith** [comment]
@swift-ci please test

**Comment by dexonsmith** [comment]
@swift-ci test please

**Comment by dexonsmith** [comment]
@swift-ci test

**Comment by dexonsmith** [comment]
@swift-ci test

**Comment by shahmishal** [comment]
@swift-ci test

**Comment by dexonsmith** [comment]
Linux passed, but it's hung on updating status: https://ci.swift.org/job/pr-apple-llvm-project-linux/591

macOS is still running: https://ci.swift.org/job/apple-llvm-project-pr-macos/560

**Comment by dexonsmith** [comment]
macOS has now passed as well.

File: llvm/include/llvm/CAS/CASNodeSchema.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include "llvm/CAS/CASReference.h"
+#include "llvm/CASObjectFormats/Data.h"
+#include "llvm/ExecutionEngine/JITLink/JITLink.h"
+#include "llvm/Support/ExtensibleRTTI.h"
+
+namespace llvm {
+namespace cas {
+
+/// NodeSchema for object hierarchies.
+///
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/CASNodeSchema.h:21]: Bonus comment line

**Review by cachemeifyoucan — APPROVED** [review]
LGTM

File: llvm/include/llvm/CAS/CASNodeSchema.h (hunk 1 above)
Comment (dexonsmith) [reply llvm/include/llvm/CAS/CASNodeSchema.h:21]: Fixed in 3c070754a30423ec21c4bb2b0f3366756c03600c.

**Comment by dexonsmith** [comment]
@benlangmuir , I'm going to land this, but can revert (or respond post-commit) if you wanted to take another look since you didn't explicitly LGTM.

**Pull request merged**